include/vcmi/events/SubscriptionRegistry.h

```diff
@@ -0,0 +1,166 @@
+/*
+ * SubscriptionRegistry.h, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#pragma once
+
+class Environment;
+
+namespace events
+{
+
+class EventBus;
+
+class DLL_LINKAGE EventSubscription : public boost::noncopyable
+{
+public:
+	virtual ~EventSubscription() = default;
+};
+
+template <typename E>
+class SubscriptionRegistry : public boost::noncopyable
+{
+public:
+	using PreHandler = std::function<void(E &)>;
+	using ExecHandler = std::function<void(E &)>;
+	using PostHandler = std::function<void(const E &)>;
+	using BusTag = const void *;
+
+	std::unique_ptr<EventSubscription> subscribeBefore(BusTag tag, PreHandler && cb)
+	{
+		boost::unique_lock<boost::shared_mutex> lock(mutex);
+
+		auto storage = std::make_shared<PreHandlerStorage>(std::move(cb));
+		preHandlers[tag].push_back(storage);
+		return make_unique<PreSubscription>(tag, storage);
+	}
+
+	std::unique_ptr<EventSubscription> subscribeAfter(BusTag tag, PostHandler && cb)
+	{
+		boost::unique_lock<boost::shared_mutex> lock(mutex);
+
+		auto storage = std::make_shared<PostHandlerStorage>(std::move(cb));
+		postHandlers[tag].push_back(storage);
+		return make_unique<PostSubscription>(tag, storage);
+	}
+
+	void executeEvent(const EventBus * bus, E & event, const ExecHandler & execHandler)
+	{
+		boost::shared_lock<boost::shared_mutex> lock(mutex);
+		{
+			auto it = preHandlers.find(bus);
+
+			if(it != std::end(preHandlers))
+			{
+				for(auto & h : it->second)
+					(*h)(event);
+			}
+		}
+
+		if(event.isEnabled())
```

Comment:
That was an idea. Register prehandler, put you modified implementation there and cancel the event. That is how WoG scripts works in most cases. If another script want to explicitly reactivate an event it could do that.